libmamba/include/mamba/core/util_string.hpp

```diff
@@ -112,86 +114,146 @@ namespace mamba
+        struct has_reserve<T, std::void_t<decltype(std::declval<T>().reserve(std::size_t()))>>
+            : std::true_type
+        {
+        };
+        template <typename T>
+        inline constexpr bool has_reserve_v = has_reserve<T>::value;
+
+        std::size_t size(const char* s);
+        std::size_t size(const wchar_t* s);
+        std::size_t size(const char c);
+        template <class T>
+        inline std::size_t size(T const& s)
         {
-            joiner(result, sep);
-            joiner(result, container[i]);
+            return s.size();
         }
-        return result;
     }
 
-    void replace_all(std::string& data, const std::string& search, const std::string& replace);
-
-    void replace_all(std::wstring& data, const std::wstring& search, const std::wstring& replace);
-
-    template <typename T>
-    auto without_duplicates(std::vector<T> values)
+    // TODO(C++20) Take an input range and return a range
+    template <typename InputIt, typename UnaryFunction, typename Value>
+    UnaryFunction join_for_each(InputIt first, InputIt last, UnaryFunction func, Value const& sep)
     {
-        const auto end_it = std::unique(values.begin(), values.end());
-        values.erase(end_it, values.end());
-        return values;
+        if (first < last)
+        {
+            func(*(first++));
+            for (; first < last; ++first)
+            {
+                func(sep);
+                func(*first);
+            }
+        }
+        return func;
     }
 
-    // Note: this function only works for non-unicode!
-    std::string to_upper(const std::string_view& input);
-    std::string to_lower(const std::string_view& input);
-
-    namespace concat_impl
+    template <class Range, class Value, class Joiner = details::PlusEqual>
+    auto join(Value const& sep, const Range& container, Joiner joiner = details::PlusEqual{}) ->
+        typename Range::value_type
     {
-        template <class T>
-        inline void concat_foreach(std::string& result, const T& rhs)
+        using Result = typename Range::value_type;
+        Result out{};
+        if constexpr (details::has_reserve_v<Result>)
         {
-            result += rhs;
+            std::size_t final_size = 0;
+            auto inc_size = [&final_size](auto const& val) { final_size += details::size(val); };
+            join_for_each(container.begin(), container.end(), inc_size, sep);
+            out.reserve(final_size);
         }
+        auto out_joiner = [&](auto&& val) { joiner(out, std::forward<decltype(val)>(val)); };
+        join_for_each(container.begin(), container.end(), out_joiner, sep);
+        return out;
+    }
 
-        template <class T, class... Rest>
-        inline void concat_foreach(std::string& result, const T& rhs, const Rest&... rest)
+    // TODO(C++20) Take an input range and return a range
+    template <typename InputIt, typename UnaryFunction, typename Value>
+    UnaryFunction join_trunc_for_each(InputIt first,
```

Comment:
I have trouble understanding this function and the following `join_trunc`, could you add some comments/documentation?

Reply:
I'll add some doc though.